### DISCUSSION TITLE
How to use zigbee-herdsman-converters and not node_modules?

### DISCUSSION THREAD

**Discussion by ok-67** [Koenkk/zigbee2mqtt#21544] [General]
I have updated the definition for a Namron thermostat plug, but I did so with the source code from the "node_modules/zigbee-herdsman-converters/devices/" directory that appears after I run "npm ci" in the zigbee2mqtt repo. The source code in the node_modules directory has several differences compared to the source code in the zigbee-herdsman-converters repo, specially when i comes to importing libraries etc.  

My updated module for the thermostat plug seems to work as expected with the modifications I made to the namron.js file, but since I made this to the file in the node_modules directory, it seems to be complicated to have it included in the  [zigbee-herdsman-converters.](https://github.com/Koenkk/zigbee-herdsman-converters.git).

So my question is, how do I set up my system to  and perform the development in the zigbee-herdsman-converters repo and have this running on my computer. 

PS, I am running zgbee2mqtt in a terminal window under Ubuntu. (npm start)

I am novice when it comes to java script and git, but have some experience with C++, so after many hours of searching through the (missing) documentation and trial and error I was able to modify the code for the Namron thermostat plug to operate as a thermostat, and I like to have this included in the official code...

**Comment by chris-1243** [comment]
To start, I am not an expert at all😉.

Why don't you use an external converter ? No need to modify the files you edited

https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html

**Reply by ok-67** [reply]
The point is that I like to contribute the code back to the source code, so that it is included in future releases of Z2M. Then it will be available for every one to use it.

**Reply by chris-1243** [reply]
I would recommend to read the link I add to my comment.

As soon as you have a device perfectly running with this external converter, it is then time to make a PR to zigbee2mqtt.io. Like that, your work will profit to the community as well😉

**Reply by sjorge** [reply]
There is a typescript -> node compile step (tsc) that will compile the .ts files in src/ into the .js files you find in node_modules.

When working on devices I usually just npm link a full zigbee-herdsman-converter repo and run `npm run build` before starting zigbee2mqtt. Sometimes npm link can be a bit finniky and not work for some reason though.

**Comment by ok-67** [comment]
Is there a description somewhere about how to do this? What files goes where etc.

**Reply by ok-67** [reply]
I have tried to follow this guide: https://docs.npmjs.com/cli/v8/commands/npm-link
I change to the directory: ....../zigbee-herdsman-converters and run the npm link command, and see there there is made a link to/usr/local/lib/node_modules
I then change to the directory where zigbee2mqtt is installed ......./zigbee2mqtt/ and run npm link modules
Then the ......./zigbee2mqtt/node_modules/ directory is populated with the original files and not the files that I have modified in  ....../zigbee-herdsman-converters

Can someone guide me in the right direction?

**Comment by Koenkk** [comment]
What you can do is, from the z2m folder:
```
cd node_modules
rm -rf zigbee-herdsman-converters
git clone https://github.com/Koenkk/zigbee-herdsman-converters.git
cd zigbee-herdsman-converters
npm ci
# Make your changes, and after every change:
npm run build
```

**Reply by ok-67** [reply]
Thank you :-)

**Reply by clockbrain** [reply]
> What you can do is, from the z2m folder:
> 
> ```
> cd node_modules
> rm -rf zigbee-herdsman-converters
> git clone https://github.com/Koenkk/zigbee-herdsman-converters.git
> cd zigbee-herdsman-converters
> npm ci
> # Make your changes, and after every change:
> npm run build
> ```

I'm trying to follow these steps with the latest zigbee2mqtt 1.40.1 but I get a compile error with zigbee-herdman-converters. Is it just a matter of bad timing and incompatible source versions or is there a different set of steps I should follow to be able to edit and compile one of the device typescript source files?

I get 12 of these sort of compile errors:

```
npm run build
> zigbee-herdsman-converters@20.17.0 build
> tsc

src/converters/toZigbee.ts:4529:7 - error TS2742: The inferred type of 'converters' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

4529 const converters = {...converters1, ...converters2, ...converters3};
           ~~~~~~~~~~

src/converters/toZigbee.ts:4529:7 - error TS2742: The inferred type of 'converters' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/tstype'. This is likely not portable. A type annotation is necessary.

4529 const converters = {...converters1, ...converters2, ...converters3};
           ~~~~~~~~~~

src/lib/ledvance.ts:29:14 - error TS2742: The inferred type of 'ledvanceTz' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

29 export const ledvanceTz = {
                ~~~~~~~~~~
```

etc.

```
Found 12 errors in 10 files.

Errors  Files
     2  src/converters/toZigbee.ts:4529
     1  src/lib/ledvance.ts:29
     1  src/lib/legacy.ts:8423
     1  src/lib/legrand.ts:118
     1  src/lib/lumi.ts:3593
     1  src/lib/ota/zigbeeOTA.ts:102
     1  src/lib/philips.ts:100
     1  src/lib/tuya.ts:1233
     2  src/lib/utils.ts:247
     1  src/lib/zosung.ts:211
```

**Reply by clockbrain** [reply]
So the solution is to not download and develop zigbee-herdsman-converters within the node_modules folder. Instead download and develop it in a separate folder such as /opt/zigbee-herdsman-converters and create a symlink to it from zigbee2mqtt/node_modules.

**Comment by devkaiwang** [comment]
I deleted `zigbee-herdsman-converters` module folder under z2m and cloned my forked repo of it. But I got error as this, please kindly help:

```log
D:\other\zigbee2mqtt\node_modules\zigbee-herdsman-converters>pnpm install
Lockfile is up to date, resolution step is skipped
Packages: +170
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 170, reused 170, downloaded 0, added 170, done

dependencies:
+ buffer-crc32 1.0.0
+ iconv-lite 0.6.3
+ semver 7.7.1
+ zigbee-herdsman 3.4.5

devDependencies:
+ @biomejs/biome 1.9.4
+ @types/buffer-crc32 0.2.4
+ @types/node 22.13.11
+ @types/semver 7.5.8
+ @vitest/coverage-v8 3.0.9
+ fast-deep-equal 3.1.3
+ husky 9.1.7
+ rimraf 6.0.1
+ ts-morph 25.0.1
+ tsc-watch 6.2.1
+ typescript 5.8.2
+ vitest 3.0.9

> zigbee-herdsman-converters@23.15.0 prepare D:\other\zigbee2mqtt\node_modules\zigbee-herdsman-converters
> husky


Done in 2.3s using pnpm v10.6.1

D:\other\zigbee2mqtt\node_modules\zigbee-herdsman-converters>pnpm run build

> zigbee-herdsman-converters@23.15.0 build D:\other\zigbee2mqtt\node_modules\zigbee-herdsman-converters
> tsc && node dist/indexer.js

src/lib/ewelink.ts:17:14 - error TS2742: The inferred type of 'ewelinkToZigbee' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

17 export const ewelinkToZigbee = {
                ~~~~~~~~~~~~~~~

src/lib/ledvance.ts:28:14 - error TS2742: The inferred type of 'ledvanceTz' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

28 export const ledvanceTz = {
                ~~~~~~~~~~

src/lib/legacy.ts:6766:7 - error TS2742: The inferred type of 'toZigbee' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

6766 const toZigbee = {...toZigbee1, ...toZigbee2};
           ~~~~~~~~

src/lib/legrand.ts:118:14 - error TS2742: The inferred type of 'tzLegrand' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

118 export const tzLegrand = {
                 ~~~~~~~~~

src/lib/lumi.ts:1423:14 - error TS2742: The inferred type of 'lumiModernExtend' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

1423 export const lumiModernExtend = {
                  ~~~~~~~~~~~~~~~~

src/lib/lumi.ts:3725:14 - error TS2742: The inferred type of 'toZigbee' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

3725 export const toZigbee = {
                  ~~~~~~~~

src/lib/namron.ts:71:14 - error TS2742: The inferred type of 'toZigbee' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

71 export const toZigbee = {
                ~~~~~~~~

src/lib/philips.ts:165:7 - error TS2742: The inferred type of 'philipsTz' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

165 const philipsTz = {
          ~~~~~~~~~

src/lib/tuya.ts:1528:7 - error TS2742: The inferred type of 'tuyaTz' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

1528 const tuyaTz = {
           ~~~~~~

src/lib/utils.ts:232:17 - error TS2742: The inferred type of 'enforceEndpoint' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

232 export function enforceEndpoint(entity: Zh.Endpoint, key: string, meta: Tz.Meta) {
                    ~~~~~~~~~~~~~~~

src/lib/utils.ts:447:17 - error TS2742: The inferred type of 'getEntityOrFirstGroupMember' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

447 export function getEntityOrFirstGroupMember(entity: Zh.Group | Zh.Endpoint) {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/lib/zosung.ts:211:14 - error TS2742: The inferred type of 'tzZosung' cannot be named without a reference to 'zigbee-herdsman-converters/node_modules/zigbee-herdsman/dist/controller/model'. This is likely not portable. A type annotation is necessary.

211 export const tzZosung = {
                 ~~~~~~~~


Found 12 errors in 10 files.

Errors  Files
     1  src/lib/ewelink.ts:17
     1  src/lib/ledvance.ts:28
     1  src/lib/legacy.ts:6766
     1  src/lib/legrand.ts:118
     2  src/lib/lumi.ts:1423
     1  src/lib/namron.ts:71
     1  src/lib/philips.ts:165
     1  src/lib/tuya.ts:1528
     2  src/lib/utils.ts:232
     1  src/lib/zosung.ts:211
 ELIFECYCLE  Command failed with exit code 1.
```

**Reply by clockbrain** [reply]
You need to move zigbee-herdsman-converters out from under zigbee2mqtt/node_modules to its own upper level folder. In your case that would be D:\other\zigbee-herdsman-converters.

However, I couldn't manage to create the necessary symlink from zigbee2mqtt/node_modules/ to that separate upper level folder for zigbee-herdsman-converters using Windows as a development environment. I expected it should work but https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/mklink didn't play nice for me.

I ended up moving my development to WSL2 (https://learn.microsoft.com/en-us/windows/wsl/install) and the Ubuntu symlink there worked as expected.

**Reply by devkaiwang** [reply]
Thank you so much for the help! Yes, I also met this issue when using symlink out of the folder of Z2M. I will switch to Linux and try again.

**Reply by devkaiwang** [reply]
Can anyone give a tutorial? I have tried natively on Windows and WSL. No luck.

**Reply by clockbrain** [reply]
Here is some guidance for developing z2m in WSL. I can't promise it is 100% complete or correct as I didn't take progressive notes when I initially set it up but these should get you most of the way.

### Caveat

USB passthrough from Windows to WSL-2 seems to have a timeout and only works for a period of time ??30 minutes?? before it drops out. Just need to reattach it with 'usbipd'
Not usable as a production environment but ok for development and testing.

### Setup

https://learn.microsoft.com/en-us/windows/wsl/install
https://learn.microsoft.com/en-us/windows/dev-environment/javascript/nodejs-on-wsl
(follow the guidance to also install Windows Terminal, nvm, node.js, and npm, VSCode and extensions, and Git)

Install usbipd-win to connect USB devices
https://learn.microsoft.com/en-us/windows/wsl/connect-usb

### Each session

Right click on Windows Terminal and open Ubuntu

Click the down arrow at the top of the WSL Ubuntu window (next to the '+'). Right click Windows PowerShell and select 'Run as administrator'
Yes to 'Do you want to allow this app to make changes to your device?'.

In PowerShell list the USB devices:
`usbipd list`
In this case my Sonoff ZBDongle-E shows as BUSID 5-1
```
Connected:
BUSID  VID:PID    DEVICE                                                        STATE
5-1    1a86:55d4  USB-Enhanced-SERIAL CH9102 (COM4)                             Not shared
```

Bind the device - might only be necessary the first time but doesn't hurt to do it each new session.
`usbipd bind --busid 5-1`
It should now show as 'Shared' in the usbipd list command.
```
Connected:
BUSID  VID:PID    DEVICE                                                        STATE
5-1    1a86:55d4  USB-Enhanced-SERIAL CH9102 (COM4)                             Shared
```

Attach the device to the WSL session.
`usbipd attach --wsl --busid 5-1`
It should now show as 'Attached' in the usbipd list command.
```
Connected:
BUSID  VID:PID    DEVICE                                                        STATE
5-1    1a86:55d4  USB-Enhanced-SERIAL CH9102 (COM4)                             Attached
```

Back to WSL-2 Ubuntu and check it shows the adapter
`lsusb`
```
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 003: ID 1a86:55d4 QinHeng Electronics SONOFF Zigbee 3.0 USB Dongle Plus V2
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
```
`ls /dev/serial/by-id/`
`usb-ITEAD_SONOFF_Zigbee_3.0_USB_Dongle_Plus_V2_20240105155310-if00`

### Developing with zigbee2mqtt

Fork https://github.com/Koenkk/zigbee2mqtt to your own GitHub account
Also fork https://github.com/Koenkk/zigbee-herdsman-converters and/or https://github.com/Koenkk/zigbee-herdsman as needed

Install zigbee2mqtt in WSL. Almost the same as https://www.zigbee2mqtt.io/guide/installation/01_linux.html but get the dev branch and get it from your forked copy in GitHub:
```
sudo mkdir /opt/zigbee2mqtt
sudo chown -R ${USER}: /opt/zigbee2mqtt
git clone --depth 1 -b dev https://github.com/MYGITHUB/zigbee2mqtt.git /opt/zigbee2mqtt
cd /opt/zigbee2mqtt
pnpm i --frozen-lockfile
```

Test that zigbee2mqtt works locally in WSL before attempting any development.

Note the developer guidance: https://github.com/Koenkk/zigbee2mqtt?tab=readme-ov-file#developing

Follow a similar process to fork and download zigbee-herdsman-converters to /opt/zigbee-herdsman-converters. Download the master branch for zigbee-herdsman-converters (there isn't a dev branch as it has much more granular versioning).

Remove zigbee-herdsman-converters from zigbee2mqtt/node_modules and link to the separate install:
```
rm -rf /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters
ln -s /opt/zigbee-herdsman-converters/ /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters
cd zigbee-herdsman-converters
```

To develop, use VSCode to edit the source files. 
Open a VSCode session:
`code .`
Use the GitHub extension in VSCode to create a new branch.

Compile and test in WSL
```
pnpm run check --fix
pnpm run build
pnpm run test
```

After testing your changes use GitHub extension in VSCode to stage and commit changes and to push to your GitHub account.

Also see: https://github.com/Koenkk/zigbee-herdsman-converters?tab=readme-ov-file#submitting-a-pull-request

When your changes are complete and working use your GitHub account to create a Pull Request against Koenkk's zigbee2mqtt or zigbee-herdsman-converters as appropriate.

**Reply by devkaiwang** [reply]
`ln` cannot work on my side, but `pnpm link` does:

```
cd ~/zigbee-herdsman-converters
pnpm link --global

cd ~/zigbee2mqtt
pnpm link zigbee-herdsman-converters
```

Additionally, I met this error when using `/dev/ttyUSB0` in the `configuration.yaml`:
```log
Error: Permission denied, cannot open /dev/ttyUSB0
please assign access to this USB port.
```
`sudo chmod 666 /dev/ttyUSB0` solves the problem.

I'm creating a PR now. Thank you @clockbrain and the community.
